File: C_-_Jumping_Takahashi.py
# receive input
N, X = input().split()
N = int(N)
X = int(X)
A = []
B = []
for i in range(N):
  a, b = input().split()
  A.append(int(a))
  B.append(int(b))

# solve
prev = [False] * (X+1)
prev[0] = True # Takahashi was initially at x = 0
for turn in range(N):
  current = [False] * (X+1)
  for x in range(X):
    if prev[x]: # if it was able to reach x in the previous turn
      if x + A[turn] <= X:
        current[x+A[turn]] = True # mark x + A[i] as possible
      if x + B[turn] <= X:
        current[x+B[turn]] = True # mark x + B[i] as possible
  prev = current

if prev[X]:
  print("Yes")
else:
  print("No")

File: C_-_Jumping_Takahashi.cpp
#include <bits/stdc++.h>
using namespace std;

void solve() {
  int N,X;
  cin >> N >> X;
  vector<int> A(N), B(N);
  for(int i=0;i<N;i++) cin >> A[i], cin >> B[i];

  vector<bool> prev(X+1);
  prev[0] = true; // Takahashi was initially at x = 0
  for(int t=0;t<N;t++){
    vector<bool> current(X+1);
    for(int x=0;x<X;x++){
      if (prev[x]) { // if it was possible to reach x in the previous turn
        if (x+A[t] <= X) 
          current[x+A[t]] = true; // mark x + A[t] as possible
        if (x+B[t] <= X)
          current[x+B[t]] = true; // makr x + B[t] as possible
      }
    }
    prev = move(current);
  }

  if (prev[X]) {
    cout << "Yes" << "\n";
  } else {
    cout << "No" << "\n";
  }
}

int main() {
  solve();

  return 0;
}
